### PR TITLE
fix cumsum op for API 2.0, optimize performance    test=develop

### PR DESCRIPTION
**Pull request by LutaoChu**
### PR types
 Function optimization, Performance optimization

### PR changes
OPs

### Describe
1 optimize forward performance
image
2 optimize backward performance
|--|优化前|优化后|结论
|-|-|-|-|
|反向耗时|313ms|0.185ms|加速1691倍

3 add parameters "dtype", "name"
4 if axis=None, flatten the input
5 full negative indexing for the  'axis' parameter is supported

### CODE REVIEW

File: python/paddle/tensor/math.py [hunk 1]
```diff
@@ -1729,3 +1728,74 @@ def kron(x, y, out=None, name=None):
         check_variable_and_dtype(out, 'out', ['float16', 'float32', 'float64', 'int32', 'int64'], 'kron')
     helper.append_op(type="kron", inputs={"X": x, "Y": y}, outputs={"Out": out})
     return out
+
+
+def cumsum(x, axis=None, dtype=None, name=None):
+    """
+	:alias_main: paddle.cumsum
+	:alias: paddle.cumsum,paddle.tensor.cumsum,paddle.tensor.math.cumsum
+	:old_api: paddle.fluid.layers.cumsum
```
Comment (wawltor) [inline python/paddle/tensor/math.py]: 删除这行

Comment (LutaoChu) [reply python/paddle/tensor/math.py]: 已删除

**Review by TCChenlong — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/cumsum_op.cc [hunk 2]
```diff
@@ -22,7 +22,14 @@ class CumOp : public framework::OperatorWithKernel {
   using framework::OperatorWithKernel::OperatorWithKernel;
 
   void InferShape(framework::InferShapeContext *ctx) const override {
-    ctx->SetOutputDim("Out", ctx->GetInputDim("X"));
+    if (ctx->Attrs().Get<bool>("flatten")) {
```
Comment (XiaoguangHu01) [inline paddle/fluid/operators/cumsum_op.cc:25]: 使用到了新增属性，这里是否会破坏兼容性？
考虑下1.8模型训练后保存inference_model，然后用2.0进行预测的场景。

File: paddle/fluid/operators/cumsum_op.cc [hunk 3]
```diff
@@ -37,6 +44,10 @@ class CumsumOpMaker : public framework::OpProtoAndCheckerMaker {
                  "dimension [default -1].")
         .SetDefault(-1)
         .EqualGreaterThan(-1);
+    AddAttr<bool>("flatten",
```
Comment (XiaoguangHu01) [inline paddle/fluid/operators/cumsum_op.cc:46]: 新增属性，是否能保持兼容？
考虑下1.8模型训练后保存inference_model，然后用2.0进行预测的场景。

File: paddle/fluid/operators/cumsum_op.cu [hunk 4]
```diff
@@ -251,34 +252,47 @@ class CumCUDAKernel : public framework::OpKernel<T> {
...
 
-    int scan_dim_size = in_dims[axis];
-    bool optimize_condition = (axis == (in_dims.size() - 1)) ? true : false;
+    T* out_data = out->mutable_data<T>(context.GetPlace());
+    const T* in_data = in->data<T>();
+    // Use thrust for parallel acceleration when the input size is equal to the
+    // size of the ‘axis’ dimension. Invalid in reverse case because the thrust
+    // APIs do not support.
+    if (size == out_dims[axis] && !reverse) {
+      if (exclusive) {
```
Comment (XiaoguangHu01) [inline paddle/fluid/operators/cumsum_op.cu]: 反向时是否有必要添加CUDA Kernel，以提升速度？

File: paddle/fluid/operators/cumsum_op.cc (hunk 2 above)
Comment (LutaoChu) [reply paddle/fluid/operators/cumsum_op.cc:25]: 考虑的场景只是1.8保存的静态图模型能不能用2.0静态图预测吗？

Comment (LutaoChu) [reply paddle/fluid/operators/cumsum_op.cc:25]: 自测不会破坏兼容性

File: paddle/fluid/operators/cumsum_op.cc (hunk 3 above)
Comment (LutaoChu) [reply paddle/fluid/operators/cumsum_op.cc:46]: 自测不会破坏兼容性

File: paddle/fluid/operators/cumsum_op.cu (hunk 4 above)
Comment (LutaoChu) [reply paddle/fluid/operators/cumsum_op.cu]: 已添加反向时的CUDA kernel，可加速1691倍

File: python/paddle/tensor/math.py [hunk 5]
```diff
@@ -1674,3 +1673,73 @@ def kron(x, y, out=None, name=None):
...
+
+def cumsum(x, axis=None, dtype=None, name=None):
+    """
+	:alias_main: paddle.cumsum
+	:alias: paddle.cumsum,paddle.tensor.cumsum,paddle.tensor.math.cumsum
+
+    The cumulative sum of the elements along a given axis. The first element of the result is the same of the first element of the input. 
+
+    Args:
+        x (Variable): Input of cumsum operator, the Tensor/LoDTensor needed to be cumsumed.
```
Comment (wawltor) [inline python/paddle/tensor/math.py]: Variable->Tensor

File: python/paddle/tensor/math.py [hunk 6]
```diff
@@ -1674,3 +1673,73 @@ def kron(x, y, out=None, name=None):
...
+    The cumulative sum of the elements along a given axis. The first element of the result is the same of the first element of the input. 
+
+    Args:
+        x (Variable): Input of cumsum operator, the Tensor/LoDTensor needed to be cumsumed. 
+        axis (int, optional): The dimension to accumulate along. -1 means the last dimension. The default (None) is to compute the cumsum over the flattened array.
+        dtype (str|np.dtype|core.VarDesc.VarType, optional): The desired data type of returned tensor. If specified, the input tensor is casted to dtype before the operation is performed. This is useful for preventing data type overflows. The default value is None. 
+        name(str, optional): Normally there is no need for user to set this property.  For more information, please refer to :ref:`api_guide_Name`. The default value is None.  
+
+    Returns:
+        Variable(Tensor/LoDTensor): The result of cumsum operator, output of cumsum operator.
```
Comment (wawltor) [inline python/paddle/tensor/math.py]: Variable->Tensor

Comment (wawltor) [reply python/paddle/tensor/math.py]: Tensor, the result of cumsum operator, output of cumsum operator.

**Comment by T8T9** [comment]
能把性能对比的 code 发出来么？谢谢~

**Comment by LutaoChu** [comment]
> 能把性能对比的 code 发出来么？谢谢~

可以的。需要运行backward=True和backward=False两种case，反向的耗时需要换算一下，把backward的耗时减去前向的耗时就是反向的耗时。

```python
import paddle
from paddle.imperative import to_variable
import numpy as np
import paddle.fluid.dygraph as dg
import paddle.fluid as fluid
import paddle.fluid.layers as F
import paddle.fluid.profiler as profiler


def test_speed(num_epochs=5, axis=0, backward=False):
    num_class = 3
    probas_data = np.random.random((2, 3, 1000, 1000))
    probas_data = probas_data.astype(np.float32)
    labels_data = np.random.randint(0, num_class, [1, 1], dtype='int32')
    
    
    probas_shape = [3, 3, 3]
    labels_shape = list(labels_data.shape)
    probas = fluid.layers.data(name='p', shape=probas_shape, dtype='float32')
    labels = fluid.layers.data(name='l', shape=labels_shape, dtype='int32')
    
    param_attr = fluid.ParamAttr(name='conv2d.weight', initializer=fluid.initializer.ConstantInitializer(value=2.0))
    y_predict = fluid.layers.conv2d(input=probas, num_filters=num_class, filter_size=2, param_attr=param_attr)
    y_predict = paddle.reshape(y_predict, shape=[-1,1])
    y_predict = paddle.cumsum(y_predict, axis=axis)
    

    labels = fluid.layers.cast(labels, dtype='float32')
    cost = fluid.layers.square_error_cost(input=y_predict, label=labels)  
    loss = fluid.layers.mean(cost)  

    if backward:
        sgd_optimizer = fluid.optimizer.SGD(learning_rate=0.1)
        sgd_optimizer.minimize(loss)

    use_cuda = True
    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace() 
    main_program = fluid.default_main_program()
    exe = fluid.Executor(place)
    exe.run(fluid.default_startup_program())
    for i in range(5):
        exe.run(
            main_program,
            fetch_list= [],
            feed={'p': probas_data, 'l': labels_data, 'm': ignore_data}, 
            return_numpy=True)

    with profiler.profiler('GPU', 'total', '/tmp/profile') as prof:
        for i in range(num_epochs):
            exe.run(
                main_program,
                fetch_list= [],
                feed={'p': probas_data, 'l': labels_data, 'm': ignore_data}, 
                return_numpy=True)
    
# test_speed(num_epochs=300)
test_speed(num_epochs=300, backward=True)
```

File: python/paddle/tensor/math.py (hunk 5 above)
Comment (LutaoChu) [reply python/paddle/tensor/math.py]: done

File: python/paddle/tensor/math.py (hunk 6 above)
Comment (LutaoChu) [reply python/paddle/tensor/math.py]: done

File: python/paddle/tensor/math.py [hunk 7]
```diff
@@ -1543,3 +1542,73 @@ def kron(x, y, name=None):
     out = helper.create_variable_for_type_inference(dtype=x.dtype)
     helper.append_op(type="kron", inputs={"X": x, "Y": y}, outputs={"Out": out})
     return out
+
+
+def cumsum(x, axis=None, dtype=None, name=None):
```
Comment (XiaoguangHu01) [inline python/paddle/tensor/math.py:1547]: 在fluid.layers.cumsum函数添加@deprecated装饰器，指明跟新API的关系。

**Review by XiaoguangHu01 — APPROVED** [review]

File: python/paddle/tensor/math.py [hunk 8]
```diff
@@ -1543,3 +1542,73 @@ def kron(x, y, name=None):
...
+    """
+	:alias_main: paddle.cumsum
+	:alias: paddle.cumsum,paddle.tensor.cumsum,paddle.tensor.math.cumsum
```
Comment (TCChenlong) [inline python/paddle/tensor/math.py:1550]: alias都不用写，可以删除这两行

**Review by TCChenlong — APPROVED** [review]
LGTM

File: python/paddle/tensor/math.py (hunk 7 above)
Comment (LutaoChu) [reply python/paddle/tensor/math.py:1547]: 已在新的PR中添加@deprecated装饰器  https://github.com/PaddlePaddle/Paddle/pull/26104
